Add hot connection pool via the minConnections setting

pjazdzewski1990 opened the pull request:
Should NOT be merged. Opened to ask code related questions
#20456: add hot connection pool via the minConnections setting
- added the akka.http.host-connection-pool.min-connections setting to reference.conf
- modified config implementation s o it's being read, with 0 as default
- based on the setting two types of pools will be created - hot and cold
- hot ones will start ASAP and not close when the number of active connections fall to 0

My first PR, so I would appreciate help and assume it's mostly wrong :)

pjazdzewski1990 committed:
20456: add hot connection pool via the minConnections setting

* added the akka.http.host-connection-pool.min-connections setting to reference.conf
* modified config implementation s o it's being read, with 0 as default
* based on the setting two types of pools will be created - hot and cold
* hot ones will start ASAP and not close when the number of active connections fall to 0

pjazdzewski1990 commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala:
```diff
@@ -228,6 +228,36 @@ class ConnectionPoolSpec extends AkkaSpec("""
+      }
+
+      // for lower bound of five connections
+      val minFiveConnections = 5
```
I'm not sure how to validate if there are exactly 5 open without modifying the code. Any suggestions?

ktoso replied:
I'd go ahead with modifying the code to support this level of introspectability.
You'll notice that the pool interface actor does actually notify the master about it's state, something among these lines and exposing the information is a good idea (we could use it for monitoring too!)

pjazdzewski1990 commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala:
```diff
@@ -228,6 +228,36 @@ class ConnectionPoolSpec extends AkkaSpec("""
+
+      condHolds(1200.millis) { () ⇒
+        Await.result(gatewayOneConnection.poolStatus(), 300.millis).get shouldBe a[PoolInterfaceRunning]
+      }
```
What I do here is I make sure the given condition holds for a period of time. If there's a better approach, please let me know.

ktoso replied:
Ok, let's have it this way for now - until we have it introspectable.
1 second is likely a bit much

ktoso commented:
Resolves https://github.com/akka/akka/issues/20456

ktoso commented:
OK TO TEST

ktoso commented on akka-http-core/src/main/resources/reference.conf:
```diff
@@ -177,6 +177,13 @@ akka.http {
     # single host endpoint is allowed to establish. Must be greater than zero.
     max-connections = 4
 
+    # The minimum number of parallel connections that a pool can establish.
```
`can establish => should keep alive ("hot")`

ktoso commented on akka-http-core/src/main/resources/reference.conf:
```diff
@@ -177,6 +177,13 @@ akka.http {
+    # If the number of connections is falling below the given threshold, new ones are being spawned.
+    # You can use this setting to build a hot pool of "always on" connections.
+    # Default is 0, meaning there might be no active connection at given moment.
+    # Keep in mind that `min-connections` should be smaller than `max-connections`
```
note: make sure we validate that with a `require()` in the code

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolFlow.scala:
```diff
@@ -77,14 +77,20 @@ private object PoolFlow {
-        .map(b.add)
+      //The difference between hot and cold slots:
+      // - hot: should be kept alive, the number is based on "minConnections" from settings
+      // - cold: hold ordinary connection that might be killed, the number is "maxConnections" - "minConnections"
```
connection  => connection

ktoso replied:
```
their number is _at most_ ` maxConnections - minConnections`
```

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/settings/ConnectionPoolSettingsImpl.scala:
```diff
@@ -21,6 +22,8 @@ private[akka] final case class ConnectionPoolSettingsImpl(
 
   require(maxConnections > 0, "max-connections must be > 0")
+  require(minConnections >= 0, "min-connections must be >= 0")
+  require(minConnections < maxConnections, "min-connections must be < max-connections")
```
`<=`?

ktoso replied:
👍

pjazdzewski1990 replied:
I will change it

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/settings/ConnectionPoolSettingsImpl.scala:
```diff
@@ -33,6 +36,7 @@ object ConnectionPoolSettingsImpl extends SettingsCompanion[ConnectionPoolSettin
   def fromSubConfig(root: Config, c: Config) = {
     ConnectionPoolSettingsImpl(
       c getInt "max-connections",
+      c getIntOpt "min-connections" getOrElse 0,
```
Why opt?

ktoso replied:
please no defaults in code – 0 will be loaded as default from config.

pjazdzewski1990 replied:
Just to confirm - will it be a compatible change? (Assuming I will uncomment the line in reference.conf) ?

ktoso replied:
It will be backwards-compatible, yes.
So you can bump the Akka version and it'll work, this is because bumping the version will also bring in the "new" reference.conf which includes the new setting.

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala:
```diff
@@ -51,7 +51,7 @@ private[http] final class PoolMasterActor extends Actor with ActorLogging {
       throw new IllegalStateException(s"pool interface actor for $gateway already exists")
     }
-    val props = Props(new PoolInterfaceActor(gateway)).withDeploy(Deploy.local)
+    val props = PoolInterfaceActor.props(gateway).withDeploy(Deploy.local)
```
thanks; inline into actorOf :)

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolFlow.scala:
```diff
@@ -77,14 +77,20 @@ private object PoolFlow {
+        .tabulate(minConnections)(PoolSlot(_, connectionFlow, settings, isHotConnection = true))
+      val coldSlots = Vector
+        .tabulate(maxConnections - minConnections)(PoolSlot(_, connectionFlow, settings, isHotConnection = false))
+      val allSlots = (hotSlots ++ coldSlots).map(b.add)
```
is ++ cheap on Vector? Can't remember

pjazdzewski1990 replied:
@2beaucoup Can you elaborate a bit more? Isn't it the point to keep the 'hot' connected even if they could be idle?

pjazdzewski1990 replied:
Good question @ktoso Here's a link on a (bigger) discussion https://issues.scala-lang.org/browse/SI-7725
I think we could make it faster by simply changing the order - 'hotSlots' will (usually) be either smaller or empty, which would allow JVM to short-circuit during the addition. 
Would this be enough to answer the concern?

ktoso replied:
Thanks for checking – https://issues.scala-lang.org/browse/SI-7725 seems to be fixed in 2.11 so we're good (at least not `n^2`). Yes, tweak the order of things

2beaucoup replied:
I think the notion of "hot" is misleading here. We just want to have "x established connections at all times". I [sketched out](https://github.com/akka/akka/issues/20456#issuecomment-222963595) in the ticket how this can be achieved. It should be the task of the conductor to manage this. The slots shouldn't need to change much.

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolInterfaceActor.scala:
```diff
@@ -147,9 +149,12 @@ private class PoolInterfaceActor(gateway: PoolGateway)(implicit fm: Materializer
       activeIdleTimeout = Some(context.system.scheduler.scheduleOnce(timeout)(gateway.shutdown()))
     }
+
+  private def shouldStopOnIdle(): Boolean =
```
good

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/util/EnhancedConfig.scala:
```diff
@@ -28,6 +28,11 @@ private[http] class EnhancedConfig(val underlying: Config) extends AnyVal {
+  def getIntOpt(path: String): Option[Int] = underlying.hasPath(path) match {
+    case true ⇒ Option(getPossiblyInfiniteInt(path))
+    case x    ⇒ None
+  }
```
this can go.

ktoso commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala:
```diff
@@ -325,24 +355,26 @@ class ConnectionPoolSpec extends AkkaSpec("""
       idleTimeout:     Duration                 = 5.seconds,
       ccSettings:      ClientConnectionSettings = ClientConnectionSettings(system)) = {
-      val settings = new ConnectionPoolSettingsImpl(maxConnections, maxRetries, maxOpenRequests, pipeliningLimit,
+      val settings = new ConnectionPoolSettingsImpl(maxConnections, minConnections, maxRetries, maxOpenRequests, pipeliningLimit,
```
settings handles well everywhere AFAICS, good 👍

pjazdzewski1990 commented on akka-http-core/src/main/scala/akka/http/impl/settings/ConnectionPoolSettingsImpl.scala:
```diff
@@ -33,6 +36,7 @@ object ConnectionPoolSettingsImpl extends SettingsCompanion[ConnectionPoolSettin
   def fromSubConfig(root: Config, c: Config) = {
     ConnectionPoolSettingsImpl(
       c getInt "max-connections",
```
One more thing - how does binary compatibility when I'm adding something new? Jenkins has failed because of that

ktoso replied:
It will be valid, the Settings classes are specifically designed such that we can safely add settings (uff that was painful back then but now we reap the benefits :-)).

You'll need to add:

```
[info] akka-http-core: found 2 potential binary incompatibilities while checking against com.typesafe.akka:akka-http-core_2.11:2.4.2  (filtered 126)
[error]  * abstract method getMinConnections()Int in class akka.http.javadsl.settings.ConnectionPoolSettings is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.javadsl.settings.ConnectionPoolSettings.getMinConnections")
[error]  * abstract method minConnections()Int in class akka.http.scaladsl.settings.ConnectionPoolSettings is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.scaladsl.settings.ConnectionPoolSettings.minConnections")
```

to `MiMa.scala` (this output is in jenkins, once you click the expand all of the log link and search for " incompatibilities "

ktoso replied:
We are allowed to add those exceptions because its internal classes

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -48,9 +48,7 @@ private object PoolSlot {
-            settings: ConnectionPoolSettings)(implicit
-    system: ActorSystem,
-                                              fm: Materializer): Graph[FanOutShape2[RequestContext, ResponseContext, RawSlotEvent], Any] =
+            settings: ConnectionPoolSettings, isHotConnection: Boolean)(implicit system: ActorSystem, fm: Materializer): Graph[FanOutShape2[RequestContext, ResponseContext, RawSlotEvent], Any] =
```
Looks like `settings` can be removed.

pjazdzewski1990 committed:
PR remarks: english, remove getIntOpt ...

Moving logic into PoolConductor WIP

pjazdzewski1990 committed:
WIP: moving minConnections logic into PoolConductor, needs more precise tests

pjazdzewski1990 commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala:
```diff
@@ -228,6 +228,45 @@ class ConnectionPoolSpec extends AkkaSpec("""
       val (Success(_), 42) = responseOut.expectNext()
     }
+
+    "never close hot connections when minConnections key is given and >0" in new TestSetup() {
```
I'm struggling with writing a proper test for `minConnections`. Kind of testability vs encapsulation battle.
After the approach change what this test should do (I think) is reach to `slotStates` and inspect whether enough slots did connect (aren't idle). Ideally we would probably go even deeper - to slots themselves (to validate if conductor statuses reflect real life). 
The problem is that `slotStates` is well hidden - internal variable of a function, inside a private class, used only to build a sub-graph in a bigger one ... I was looking for a way to expose it, but my approach felt ugly and left a trail of `Internal API, only for testing` comments. 
Could anyone point me to an existing piece of code doing something similar?
cc: @ktoso @2beaucoup

2beaucoup replied:
This would be rather a black box test than looking at the slot internals:

1. open connection pool with `minConnections = n`
2. `acceptIncomingConnection()` `n` times
3. send `n` requests with `Connection: close` at once
4. expect `n` responses
5. `acceptIncomingConnection()` `n` times

pjazdzewski1990 replied:
That would simplify the code a lot, but I wonder what would be the assumption here? `acceptIncomingConnection()` didn't throw? This outline seems straightforward, but I'm not sure if this is enough to validate that the internal state is correct ... What do you think?

2beaucoup replied:
> That would simplify the code a lot, but I wonder what would be the assumption here? acceptIncomingConnection() didn't throw?

Yes. It would otherwise throw an `AssertionError`. Take a look at `acceptIncomingConnection`s impl and the current usages to get an idea of what it's doing.

> This outline seems straightforward, but I'm not sure if this is enough to validate that the internal state is correct ...

This way the initial eager connection and the automatic re-connections are checked. What do you think is missing?

2beaucoup replied:
Additionally you can override `testServerHandler` to include the connection counter in the response and finally send some more requests to see if the counters match the expected values.

pjazdzewski1990 replied:
Looks like the basic plan gets the stuff done :) Thanks a lot
I will try the `testServerHandler`approach too

pjazdzewski1990 committed:
added testing via expectNextN in ConnectionPoolSpec.scala; fixed the bug with skipping the actor state in PoolSlot

pjazdzewski1990 commented:
@2beaucoup @ktoso I think this one is ready for review. Do you want to discuss it in this PR or should I open a new one without WIP status? 
I've followed the general direction (`PoolConductor` decides which slots to connect), but applied some changes on my own (few things came out during testing).

ktoso commented:
Let's discuss here :)
Hope to get to reviewing it today, need to somehow finish my ScalaDays talk too today... :-)

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -38,18 +38,22 @@ private object PoolConductor {
     +--------->|   Merge   +---->| Selector  +-------------->| (MapConcat) +---->|  (Flexi   +-------------->
                |           |     |           |               |             |     |   Route)  +-------------->
-               +----+------+     +-----+-----+               +-------------+     +-----------+       to slots
+               +----+------+     +-----+-----+               +-------------+     +-----------+       to slotSettings
```
why the rename?

pjazdzewski1990 replied:
I guess IDEA might have changed it in when I was doing auto-rename. Will bring the old name back

ktoso replied:
note to self, was renamed back; good

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -63,17 +67,18 @@ private object PoolConductor {
 
   */
-  def apply(slotCount: Int, pipeliningLimit: Int, log: LoggingAdapter): Graph[Ports, Any] =
+  def apply(slotSettings: PoolSlotsSetting, pipeliningLimit: Int, log: LoggingAdapter): Graph[Ports, Any] =
```
Is the extra class necessary? Maybe use `Range`?

pjazdzewski1990 replied:
Without the extra class the function would look like this `apply(Int, Int, Int, LoggingAdapter)`. IMO Those 3 `Int`s make it easy to make a mistake. A mistake that won't be caught on compile time.
When it comes to naming - I feel like `Range` suggests that it holds some data (like the built-in one). Where in this case it's more like it hold two extremes, no the data in-between. But I'm not going to fight over it - if you prefer `Range` I will rename.

ktoso replied:
Class sounds ok

2beaucoup replied:
I'd rather drop the extra class.

pjazdzewski1990 replied:
Do we vote in this case or what? :) I don't have any strong opinions here. 
I would prefer having this additional safety net, but I understand that it's probably the only purpose of `PoolSlotsSetting`

ktoso replied:
Let's go with no extra class 👍

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -85,9 +90,12 @@ private object PoolConductor {
+  sealed trait SlotCommand
+  case class DispatchCommand(rc: RequestContext) extends SlotCommand
+  case class SwitchSlotCommand(cmd: DispatchCommand, slotIx: Int) extends SlotCommand
+  case class SlotShouldConnectCommand(slotIx: Int) extends SlotCommand
```
This is confusing. Please look at my previous example where `SwitchCommand` wraps `SlotCommand`. That would simplify the router logic which just has to unwrap.

pjazdzewski1990 replied:
If I understand correctly, You would change
`SwitchSlotCommand(cmd: DispatchCommand, slotIx: Int)` into `SwitchSlotCommand(cmd: SlotCommand, slotIx: Int)`, right? 
The reason why I didn't go that way was that it, makes (for instance) embedding a `SwitchSlotCommand` in other `SwitchSlotCommand` possible. It's valid from the compilers point of view, but it makes little sense from the logical standpoint. 
What do you think?

2beaucoup replied:
`SwitchSlotCommand` shouldn't extend anything. ;)

pjazdzewski1990 replied:
Ok, got it. Will extract switch from other commands

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -153,6 +164,27 @@ private object PoolConductor {
+            .filter(_._1 == Unconnected)
+            .take(slotSettings.minSlots - connectedNum)
+          unconnected.map(_._2).map(connect)
+        }
```
See https://github.com/2beaucoup/akka/commit/1c89d705db5fc9ada290f4b0b7bedbe7101c066b#diff-0dc3503e2c7e4c93cd4b99e13e236cacR143 how this can be simplified.

pjazdzewski1990 replied:
I will double-check if `reconnectIfNeeded()` should only always connect one slot. If it's the case I will go with`indexWhere` as you suggested. Thanks

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -205,11 +237,11 @@ private object PoolConductor {
     override def initialAttributes = Attributes.name("PoolConductor.Route")
 
-    override val shape = new UniformFanOutShape[SwitchCommand, RequestContext](slotCount)
+    override val shape = new UniformFanOutShape[SlotCommand, SlotCommand](slotCount)
```
should be `UniformFanOutShape[SwitchCommand, SlotCommand]`

pjazdzewski1990 replied:
Will do

pjazdzewski1990 committed:
fix comments from PR: make sure PoolConductor produces only SwitchSlotCommand

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -85,9 +90,12 @@ private object PoolConductor {
+  case class SlotShouldConnectCommand(slotIx: Int) extends SlotCommand
 
-  // the SlotSelector keeps the state of all slots as instances of this ADT
+  // the SlotSelector keeps the state of all slotSettings as instances of this ADT
```
here too (maybe other places)

pjazdzewski1990 committed:
fix typo: slotSettings -> slots

pjazdzewski1990 commented:
What with this PR?

ktoso commented:
You know it was ScalaDays and eng meeting @pjazdzewski1990 :-)
Back to reviewing things now :-)

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -41,6 +41,10 @@ private object PoolConductor {
         outlets.asInstanceOf[immutable.Seq[Outlet[RequestContext]]])
   }
 
+  case class PoolSlotsSetting(minSlots: Int, maxSlots: Int){
```
`final case class`

2beaucoup commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala:
```diff
@@ -228,6 +229,56 @@ class ConnectionPoolSpec extends AkkaSpec("""
+
+      // for lower bound of five connections
+      val minFiveConnections = 5
+      val (requestInFive, requestOutFive, responseOutSubFive, hcpMinFiveConnection) = cachedHostConnectionPool[Int](
```
Could you please drop the "Five" on all variables?

2beaucoup commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala:
```diff
@@ -228,6 +229,56 @@ class ConnectionPoolSpec extends AkkaSpec("""
+      val (requestInFive, requestOutFive, responseOutSubFive, hcpMinFiveConnection) = cachedHostConnectionPool[Int](
+        idleTimeout = 100.millis,
+        minConnections = minFiveConnections,
+        maxConnections = minFiveConnections + 10)
```
Why `+ 10`? These additional connections are never used.

pjazdzewski1990 replied:
It could be any number >0. The default for tests is 3, which would cause an error (min > max)

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -23,7 +23,7 @@ private object PoolConductor {
     requestIn:   Inlet[RequestContext],
     slotEventIn: Inlet[RawSlotEvent],
-    slotOuts:    immutable.Seq[Outlet[RequestContext]]) extends Shape {
+    slotOuts:    immutable.Seq[Outlet[SlotCommand]]) extends Shape {
```
👍

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -89,9 +89,10 @@ private object PoolConductor {
+  sealed trait SlotCommand
+  case class DispatchCommand(rc: RequestContext) extends SlotCommand
+  case class SwitchSlotCommand(cmd: DispatchCommand, slotIx: Int) extends SlotCommand
+  case class SlotShouldConnectCommand(slotIx: Int) extends SlotCommand
```
`final` on all case classes

2beaucoup commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala:
```diff
@@ -228,6 +229,56 @@ class ConnectionPoolSpec extends AkkaSpec("""
+      condHolds(1000.millis) { () ⇒
+        val status = gatewayFiveConnections.poolStatus()
+        Await.result(status, 100.millis).get shouldBe a[PoolInterfaceRunning]
+      }
```
Why not just another 5 `acceptIncomingConnection()`?

pjazdzewski1990 replied:
I've got a problem with this one. Is, in this context, `acceptIncomingConnection()` the same as `responseOutSub.request(1); requestOut.expectNextN(1)`? The first option doesn't work, but I'm not sure if it requests on the right subscriptions

2beaucoup commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/HttpConfigurationSpec.scala:
```diff
@@ -129,6 +129,34 @@ class HttpConfigurationSpec extends AkkaSpec {
+          akka.http.host-connection-pool.max-connections = 1
+        """.stripMargin) { sys ⇒
+
+        intercept[IllegalArgumentException] { ConnectionPoolSettings(sys) }
```
Do we explicitly test default values and requirements for other configs?

ktoso replied:
Yeah, in some HTTP things I believe we do test requirements

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -225,6 +257,7 @@ private object PoolSlot {
         context.stop(self)
     }
   }
+  def flowInportProps(s: ActorRef) = Props(new FlowInportActor(s)).withDeploy(Deploy.local)
```
:+1:

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -96,6 +101,17 @@ private object PoolSlot {
     override def requestStrategy = ZeroRequestStrategy
+
+    /**
+     * How PoolProcessor changes it's `receive`:
```
`its` because possesive

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -153,6 +165,23 @@ private object PoolConductor {
+      }
+
+      def startMinConnections(): Unit =
+        (0 until slotSettings.minSlots).foreach { connect }
```
inline?

2beaucoup commented on project/MiMa.scala:
```diff
@@ -890,7 +891,20 @@ object MiMa extends AutoPlugin {
+        ProblemFilters.exclude[MissingClassProblem]("akka.http.impl.engine.client.PoolConductor$SwitchCommand"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.impl.settings.ConnectionPoolSettingsImpl.apply"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.impl.settings.ConnectionPoolSettingsImpl.copy"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.impl.settings.ConnectionPoolSettingsImpl.this")
```
FilterAnyProblemStartingWith("akka.stream.impl")

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -153,6 +165,23 @@ private object PoolConductor {
+        (0 until slotSettings.minSlots).foreach { connect }
+
+      // pool ordering connection doesn't change the status of a connection to Idle,
+      // receiving confirmation from the Slot does. See: handler for "slotIn"
```
Why is this needed?

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -153,6 +165,23 @@ private object PoolConductor {
+        emit(out, SwitchSlotCommand(SlotShouldConnectCommand, index))
+
+      private def reconnectIfNeeded(): Unit = {
+        val connectedNum = slotStates.count(_ != Unconnected)
```
inline and remove all curly braces?

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -137,6 +146,9 @@ private object PoolConductor {
               slotStates(slotIx) = slotStateAfterDisconnect(slotStates(slotIx), failed)
+              reconnectIfNeeded()
+            case SlotEvent.ConnectedEagerly(slotIx) ⇒
+              slotStates(slotIx) = Idle
```
Doesn't this prevent requests from dispatch on this slot between sending `SlotShouldConnectCommand` and receiving  `ConnectedEagerly`? Previously dispatch of the first requests instantly set the slot state to `Loaded`/`Busy`. Maybe it should become `Idle` right away.

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -85,7 +90,11 @@ private object PoolConductor {
-  private case class SwitchCommand(rc: RequestContext, slotIx: Int)
+  sealed trait SlotCommand
+  case class DispatchCommand(rc: RequestContext) extends SlotCommand
+  case object SlotShouldConnectCommand extends SlotCommand
```
Simply `ConnectCommand`?

pjazdzewski1990 replied:
Maybe `ConnectEagerlyCommand`? It would fit well with `SlotEvent.ConnectedEagerly`

ktoso commented:
Seems like last round of comments here, cc @pjazdzewski1990

pjazdzewski1990 committed:
Fix style after code review
- inline short snippets
- shorter Mima
- better naming
- make SlotCommands final

drewhk commented:
LGTM on my side. Pretty nifty feature, I like it.

ktoso commented:
LGTM, thanks @pjazdzewski1990 !

ktoso commented:
Would you be able to send in a small PR next that adds a notice about this feature in docs?
I'll merge this so we can keep moving :-)

pjazdzewski1990 commented:
My first (real) PR merged 🎉 Thanks guys!

ktoso commented:
And it's an actual proper big feature :-) Grats and keep them coming!

patriknw commented:
Congrats and big thanks, @pjazdzewski1990
